### ISSUE TITLE
Local Audio Transcription

### ISSUE THREAD

**Issue by rafael844** [sepinf-inc/IPED#248]
Is there any intention in use some local audio transcription application?
Such as cmusphinx / Vosk ?
https://cmusphinx.github.io/
https://github.com/alphacep/vosk-api

**Comment by lfcnassif** [comment]
Yes, we plan to add an implementation using Deep Speech: https://github.com/mozilla/DeepSpeech

But we want to train a Portuguese model first, that will need some work...

**Comment by rafael844** [comment]
Thank you.

**Comment by lfcnassif** [comment]
Let's leave open to be tracked.

**Comment by rafael844** [comment]
There is Aldo an wav2letter api, from Facebook.
https://github.com/facebookresearch/wav2letter

**Comment by lfcnassif** [comment]
I didn't see vosk already has a portuguese model and java binding. Has anyone tested that model or other language accuracy?

**Comment by hauck-jvsh** [comment]
Link present in the Vosk page to a project to train portuguese modelos.
https://github.com/falabrasil/kaldi-br

**Comment by lfcnassif** [comment]
Vosk project published a new big 1.6GB portuguese model and also accuracy numbers for this new and the old model, unfortunately WER are very low:
![image](https://user-images.githubusercontent.com/7276994/176084148-ff6bd878-d122-46a4-8954-286dcebf1103.png)

That also suggests the accuracy issue is with the trained pt-BR model, not with the algorithm, since english numbers seems much better (of course, tested on different datasets):
![image](https://user-images.githubusercontent.com/7276994/176084337-32762959-74af-4b15-8b30-5f2221a1cf53.png)

**Comment by lfcnassif** [comment]
> But the new (bigger) model accuracy is better than the one we had before, right?

Yes! But probably slower too...

**Comment by rotivrotiv** [comment]
> O projeto Vosk publicou um novo grande modelo português de 1,6 GB e também **números de precisão** para este novo e o antigo modelo, infelizmente o WER é muito baixo (como pensávamos):
> 
> ![imagem](https://user-images.githubusercontent.com/7276994/176084148-ff6bd878-d122-46a4-8954-286dcebf1103.png)
> 
> Isso também sugere que o problema de precisão está relacionado ao modelo pt-BR treinado, não ao algoritmo, já que os números em inglês parecem muito melhores (claro, testados em diferentes conjuntos de dados): ![imagem](https://user-images.githubusercontent.com/7276994/176084337-32762959-74af-4b15-8b30-5f2221a1cf53.png)

Amigo, como posso usar este arquivo de 1,6 GB no IPED? Estou tentando, porém, vejo que o IPED está utilizando a versão small (para android?) e quando baixo o arquivo maior ele vem com pastas diferentes, apesar de conter quase os mesmos arquivos.

Obrigaod!

**Comment by lfcnassif** [comment]
I think you just have to replace the portuguese model into iped-4.0.0/models/vosk folder for the new one, but I have not tested yet. I can test after I return from vacation next week.

**Comment by lfcnassif** [comment]
> I think you just have to replace the portuguese model into iped-4.0.0/models/vosk/pt-BR folder for the new one, but I have not tested yet. I can test after I return from vacation next week.

Just tested the big 1.6GB model on a small 300 audios dataset. After replacing the model folder contents, you also must remove the "rescore" folder (not present in the english model, so I guessed it wasn't needed). I'm not sure about the side effects, but this fixes a `java.lang.Error: Invalid memory access` while loading the model with current used vosk-0.3.32 version and a `java.io.IOException: Failed to create a model` with vosk-0.3.38

The good news is that the huge model took that same time as the small model to transcribe that data set: 78s on a 48 threads CPU.

**Comment by lfcnassif** [comment]
> The good news is that the huge model took the same time as the small model to transcribe that data set: 78s on a 48 threads CPU.

That is wrong, my fault. Actually the big model is, surprisingly, a bit **faster** than the small model:
- vosk-0.3.32: small model took 95s - big model took 71s
- vosk-0.3.38: small model took 126s - big model took 79s

I'll run on a larger data set tomorrow to confirm that observation.

**Comment by lfcnassif** [comment]
Running time with a ~1700 audios data set:
- vosk-0.32.3: small model took 1359s - big model took 1066s
- vosk-0.3.38: small model took 2081s - big model took 1692s

But, a manual and informal accuracy comparison made by a colleague of mine seems to show the big model has worse results than the small model with some real case audios. Maybe the big model has a higher bias towards the data sets used for training, and generalize worse to different data sets, maybe...

**Comment by andrenatal** [comment]
Edit: disregard my invalid comment: I just noticed I'm talking about a different model.

**Comment by lfcnassif** [comment]
Great paper from 1 year ago, about a wav2vec-2.0 trained model for pt-BR and references to 470 hours of Portuguese data sets!
https://arxiv.org/abs/2107.11414

**Comment by lfcnassif** [comment]
Excerpt from their conclusion:

```
On average, our
model obtained 10.5% and 12.4% of WER against the proposed test sets, with
and without a language model, respectively. According to our knowledge, the
model achieves state-of-the-art results among the open available E2E models for
the target language.
```

**Comment by lfcnassif** [comment]
Author's repo:
https://github.com/lucasgris/wav2vec4bp

Seems they made models public available :-)

**Comment by lfcnassif** [comment]
Opened #1214 for those interested.
